Fatal error C1001 (Internal compiler error) when compiling with VS 2017

When updating the AppVeyor configuration an error popped up in the compile log:
```
C:\projects\alicevision\src\aliceVision/prettyprint.hpp(62): fatal error C1001: An internal error has occurred in the compiler. [C:\projects\alicevision\build\src\aliceVision\system\aliceVision_system.vcxproj]
  (compiler file 'd:\agent\_work\8\s\src\vctools\compiler\cxxfe\sl\p1\c\types.c', line 4563)
   To work around this problem, try simplifying or changing the program near the locations listed above.
```

Here the [AppVeyor configuration used](https://github.com/EwoutH/AliceVision/blob/patch-1/appveyor.yml) and the [run itself](https://ci.appveyor.com/project/AliceVision/alicevision/builds/32552550).

Blocks #775.

--- a/src/aliceVision/prettyprint.hpp
+++ b/src/aliceVision/prettyprint.hpp
@@ -49,17 +49,33 @@ namespace pretty_print
         struct has_begin_end : private sfinae_base
         {
         private:
+#ifdef _MSC_VER
+            // Work around MSVC ICE in 15.9.x by moving decltype out to template
+            // typename
+            template <typename C, typename LEFT = C::const_iterator(C::*)() const>
+            static yes & f(typename std::enable_if<
+                std::is_same<decltype(static_cast<typename LEFT>(&C::begin)),
+                             typename C::const_iterator(C::*)() const>::value>::type *);
+#else // _MSCV_VER
             template <typename C>
             static yes & f(typename std::enable_if<
                 std::is_same<decltype(static_cast<typename C::const_iterator(C::*)() const>(&C::begin)),
                              typename C::const_iterator(C::*)() const>::value>::type *);
+#endif // _MSCV_VER
 
             template <typename C> static no & f(...);
 
+#ifdef _MSC_VER
+            template <typename C, typename LEFT = C::const_iterator(C::*)() const>
+            static yes & g(typename std::enable_if<
+                 std::is_same<decltype(static_cast<typename LEFT>(&C::end)),
+                              typename C::const_iterator(C::*)() const>::value, void>::type*);
+#else // _MSCV_VER
             template <typename C>
             static yes & g(typename std::enable_if<
-                std::is_same<decltype(static_cast<typename C::const_iterator(C::*)() const>(&C::end)),
-                             typename C::const_iterator(C::*)() const>::value, void>::type*);
+                 std::is_same<decltype(static_cast<typename C::const_iterator(C::*)() const>(&C::end)),
+                              typename C::const_iterator(C::*)() const>::value, void>::type*);
+#endif // _MSCV_VER
 
             template <typename C> static no & g(...);
 

